### PR TITLE
daemon: allow directory creation in /run/secrets

### PR DESCRIPTION
**Pull request by cyphar**
Since FileMode can have the directory bit set, allow a SecretStore
implementation to return secrets that are actually directories. This is
useful for creating directories and subdirectories of secrets.

[Yoda, the cat by dia_n](https://flic.kr/p/f7aMri)

Signed-off-by: Aleksa Sarai <asarai@suse.de>

###### [Yoda, the cat](https://flic.kr/p/f7aMri) by [dia_n](https://www.flickr.com/photos/30967075@N08/), on Flickr.

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Please elaborate the detail of your plan about directory-style secret?
How API/CLI will be changed?

**Comment by cyphar** [comment]
The API and CLI haven't been changed, this is just opening the door to them being changed at a later stage (at SUSE we want to pass `credentials.d` which is a directory to Docker containers -- which is not possible with the current implementation without losing the ability to change the mode of the directory itself).

It doesn't make much sense to me that the `swarmapi` uses `os.FileMode` but there's no restriction against it being a directory (but the code always treats it as a file).

**Comment by thaJeztah** [comment]
ping @cyli @diogomonica PTAL


@cyphar if agreed upon, would it make sense to also implement the storage part? (not sure how that would work, perhaps `tar` a directory, and add it as secret, then untar when mounting in the container?

**Comment by ehazlett** [comment]
@cyphar one of the intentions to restrict it to a single directory originally was to prevent confusion about the target and masking of root level directories (i.e. `target=/bin` or something).  However, the more I think about it, we could probably support it starting from `/var/run` as long as we make it clear in the docs that it's relative to that starting path.  So perhaps something like `source=cert-2017.pem,target=certs/default.pem` might be an option.  Would that solve the `credentials.d` scenario?

**Comment by diogomonica** [comment]
We discussed supporting directories in the original design. Not supporting them was a design choice.

I haven't seen a lot of applications out there that expect a particular directory structure, so I think supporting it isn't worth taking away the simplicity of the "given a secret-target, you'll always find it in `/var/run/secret-target`".

Just my $0.02

**Comment by thaJeztah** [comment]
I'm not familiar with `credentials.d`, so can't say much about that; do those credentials _have_ to be passed "all or nothing" or do they make sense individually as well?

Slightly orthogonal, but I think having support for different backing stores would be nice

**Comment by cyphar** [comment]
@diogomonica 

> I haven't seen a lot of applications out there that expect a particular directory structure,

`SUSEConnect` is one pertinent example. Credentials for different modules and sources are stored in separate files in `/etc/zypp/credentials.d`. We already have support in `container-SUSEConnect` to be able to handle `/run/secrets/credentials.d/*` as a source for those credentials (without copying them) and we carry patches to pass `credentials.d` to containers from the host. This is necessary in order for you to be able to run SLE inside containers (and be able to install packages).

Now, we are still carrying our own patches (similar to Red Hat) to make this work, but with the advent of secrets being supported in the Docker daemon maybe we should be revisiting the old proposals to allow vendors to inject secrets into containers. Secrets used to be out of the question, but now they're implemented. Maybe we can come to an agreement on this topic.

> Do those credentials have to be passed "all or nothing" or do they make sense individually as well?

While they do make sense individually, it would require us to modify SUSEConnect and a few other tools (and I'm sure Red Hat would have to do the exact same thing) just because creating a directory was seen as not making sense. Not to mention that most users using containers on a SUSE system would like to access the same set of modules as their host system (thus effectively `credentials.d` should be passed as a single "thing").

I mean, all of this discussion is ignoring that the current swarm mode of secret handling doesn't work for a single Docker daemon. But I really doubt we'll come to an agreement on that topic.

**Comment by cpuguy83** [comment]
I definitely don't like this implementation, but being able to namespace a secret under a directory structure (within `/run/secrets` does seem neccessary for not only what @cyphar mentioned but also just in general.

**Comment by ehazlett** [comment]
It looks like I missed the original question and intent.  I was thinking you wanted to add _individual_ credentials to a directory, not a wholesale bunch of secrets.

> Credentials for different modules and sources are stored in separate files in /etc/zypp/credentials.d. We already have support in container-SUSEConnect to be able to handle /run/secrets/credentials.d/*

Sorry to go a bit off topic but I agree with what @thaJeztah said.  First, could they individually be added and if so how about just referencing `/run/secrets/*`?  I know that's redesigning but the secrets were designed with a 1:1 in mind (and we should fix the directory mode issue).

As an aside, we discussed at once point "secret groups".  I think that would be more in line with this as opposed to overriding the secrets to inject them.

**Comment by thaJeztah** [comment]
Trying to get the use case more clear (thanks for adding the additional information!); are these credentials tied to each individual host? When working with a swarm, would those credentials have to be present on each host and mounted in the container, or would it be possible that they are provided by a manager / stored in swarm, and distributed to each host (like the current secrets are distributed)?

**Comment by cyphar** [comment]
@ehazlett 

> Sorry to go a bit off topic but I agree with what @thaJeztah said. First, could they individually be added and if so how about just referencing /run/secrets/*?

We also have `/etc/SUSEConnect` which is also required. I mean, ultimately we could do something like `/run/secrets/credentials.d-*` where you effectively flatten the directory but this just smells of a hack to me.

> are these credentials tied to each individual host?

Yes, it's based on SUSE's subscription model (at least on normal SLES, CaaSP will be different but that uses Kubernetes and not swarm). Distributing it doesn't make sense. Which I know goes in the face of how secrets work now, but the reason for this PR is so that we can upstream as much of the SUSE secrets patches as possible. I'm hoping that one day it won't be necessary to carry patches in order for Docker to work with how our distributions operate.

> provided by a manager / stored in swarm

While it might be possible for us to do something like this in the future, you'd effectively have two copies of `/etc/zypp/credentials.d` floating around for every host. Not exactly ideal, especially when it comes to synchronising the copies (we have SUSE Manager that allows you to manage such clusters of machines, would we have to now add support to SUSE Manager to update the manager holding `/etc/zypp/credentials.d` as well?).

**Comment by thaJeztah** [comment]
Right, so this needs a distinction between "globally" managed secrets ("I create a service and pass it some secret. Wherever those tasks land, they need to get access to that secret"), "local" secrets (the node itself provides the secret), and "external" secrets (usecase; I have a license manager that can issue "secrets". The manager fetches the secrets, and passes it on to the task/container)

Thinking of 

    --secret provider=<default|local|external>,store=my-secrets-helper,src=my-secret,target=foo

The "provider" for "local" and "external" could be a helper application on the node that allows docker to obtain the secrets-data.

Not a security expert, so hit me if it's nonsense :smile:

**Comment by runcom** [comment]
shouldn't you lift this restriction here https://github.com/runcom/docker/commit/dcc0a71f5d29d8850f23124d098705ea3a5c2058#diff-0f20873a38571444bac38770160648a5R180 ? otherwise, directories will be blocked at that `if`

**Comment by cyli** [comment]
There has been some discussion in https://github.com/docker/docker/pull/32336 to allow targets outside of `/run/secrets`.  In that case, possibly secrets can go into `/etc/zypp/credentials.d` instead - would creating directories in `/run/secrets` still be necessary if we allowed arbitrary paths?

File: daemon/container_operations_unix.go [hunk 1]
```diff
@@ -196,8 +191,14 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
 		if secret == nil {
 			return fmt.Errorf("unable to get secret from secret store")
 		}
-		if err := ioutil.WriteFile(fPath, secret.Spec.Data, s.File.Mode); err != nil {
-			return errors.Wrap(err, "error injecting secret")
+		if s.File.Mode.IsDir() {
+			if err := os.Mkdir(fPath, s.File.Mode); err != nil {
```
Comment (runcom) [inline daemon/container_operations_unix.go]: this should be `MkdirAll` or starting a contianer, stopping it and restarting it will fail.

**Comment by AkihiroSuda** [comment]
What's current status?

**Comment by cyphar** [comment]
Since we need this for SLE, we currently apply this patch for our packages. I still think this would be useful -- but the implementation of this has drastically changed since I opened this PR (now it directory entries store a `tar` archive of the directory contents -- in order to handle some bugs that I found when using this patch).

I can push this again, but if it's not going to ever be merged then I don't really see the point of keeping this open. Again, this is something that we really need (and I think "secret plugins" would be useful -- because currently secrets are locked away behind swarm) but if there's no interest to merge it then I don't mind maintaining it out-of-tree.

**Comment by olljanat** [comment]
@cyphar this one needs to be rebased

**Comment by olljanat** [comment]
Derek add label: status/failing-ci

**Comment by cpuguy83** [comment]
Let's close since this has gone stale.
This is not an outright rejection just cleaning things up since this is not actively worked on and there is no consensus.

**Pull request closed without merging**